mono/utils/mono-os-mutex.h

```diff
@@ -310,51 +310,62 @@ WINBASEAPI BOOL WINAPI InitializeCriticalSectionEx(LPCRITICAL_SECTION lpCritical
 
 #endif /* defined __MINGW32__ && !defined __MINGW64_VERSION_MAJOR && (_WIN32_WINNT >= 0x0600) */
 
-typedef CRITICAL_SECTION mono_mutex_t;
+// FIXME mono_mutex_t and mono_mutex_recursive_t.
+typedef struct mono_mutex_t {
+	union {
+		CRITICAL_SECTION critical_section;
+		SRWLOCK srwlock;
+	};
+	gboolean recursive;
+} mono_mutex_t;
+
 typedef CONDITION_VARIABLE mono_cond_t;
 
 static inline void
 mono_os_mutex_init (mono_mutex_t *mutex)
 {
-	BOOL res;
-
-	res = InitializeCriticalSectionEx (mutex, 0, CRITICAL_SECTION_NO_DEBUG_INFO);
-	if (G_UNLIKELY (res == 0))
-		g_error ("%s: InitializeCriticalSectionEx failed with error %d", __func__, GetLastError ());
+	mutex->recursive = FALSE;
+	InitializeSRWLock (&mutex->srwlock);
 }
 
 static inline void
 mono_os_mutex_init_recursive (mono_mutex_t *mutex)
 {
-	BOOL res;
+	mutex->recursive = TRUE;
+	BOOL res = InitializeCriticalSectionEx (&mutex->critical_section, 0, CRITICAL_SECTION_NO_DEBUG_INFO);
 
-	res = InitializeCriticalSectionEx (mutex, 0, CRITICAL_SECTION_NO_DEBUG_INFO);
-	if (G_UNLIKELY (res == 0))
-		g_error ("%s: InitializeCriticalSectionEx failed with error %d", __func__, GetLastError ());
+	g_assertf (res, "%s: InitializeCriticalSectionEx failed with error %d", __func__, GetLastError ());
 }
 
 static inline void
 mono_os_mutex_destroy (mono_mutex_t *mutex)
 {
-	DeleteCriticalSection (mutex);
+	if (mutex->recursive)
```

Comment:
Maybe add a comment that SRW's doesn't need to be deleted, just for clarity indicating that this case has not been overlooked/missed (since it doesn't need to be handled)?

Reply:
Similar to what we have in mono_os_cond_destroy.

Reply:
There is no delete function. It is impossible to do otherwise..

Reply:
It is impossible to do otherwise, not just no need. but ok I'll add comment.